src/poetry/console/commands/self/self_command.py

```diff
@@ -0,0 +1,127 @@
+from typing import TYPE_CHECKING
+
+from poetry.core.packages.dependency import Dependency
+from poetry.core.pyproject.toml import PyProjectTOML
+
+from poetry.__version__ import __version__
+from poetry.console.commands.installer_command import InstallerCommand
+from poetry.factory import Factory
+from poetry.packages.project_package import ProjectPackage
+from poetry.utils.env import EnvManager
+from poetry.utils.env import SystemEnv
+from poetry.utils.helpers import directory
+
+
+if TYPE_CHECKING:
+    from poetry.poetry import Poetry
+    from poetry.utils.env import Env
+
+
+class SelfCommand(InstallerCommand):
+    ADDITIONAL_PACKAGE_GROUP = "additional"
+
+    @staticmethod
+    def get_default_system_pyproject_file() -> Path:
+        # We separate this out to avoid unwanted side effect during testing while
+        # maintaining dynamic use in help text.
+        #
+        # This is not ideal, but is the simplest solution for now.
+        from poetry.locations import CONFIG_DIR
+
+        return Path(CONFIG_DIR).joinpath("pyproject.toml")
+
+    @property
+    def system_pyproject(self) -> Path:
+        file = self.get_default_system_pyproject_file()
+        file.parent.mkdir(parents=True, exist_ok=True)
+        return file
+
+    def reset_env(self) -> None:
+        self._env = EnvManager.get_system_env(naive=True)
+
+    @property
+    def env(self) -> Env:
+        if self._env is None or not isinstance(self._env, SystemEnv):
+            self.reset_env()
+        return self._env
+
+    @property
+    def default_group(self) -> str:
+        return self.ADDITIONAL_PACKAGE_GROUP
+
+    @property
+    def activated_groups(self) -> set[str]:
+        return {self.default_group}
+
+    def generate_system_pyproject(self) -> None:
+        preserved = {}
+
+        if self.system_pyproject.exists():
+            content = PyProjectTOML(self.system_pyproject).poetry_config
+
+            for key in {"group", "source"}:
+                if key in content:
+                    preserved[key] = content[key]
+
+        package = ProjectPackage(name="poetry-instance", version=__version__)
+        package.add_dependency(Dependency(name="poetry", constraint=f"^{__version__}"))
+
+        package.python_versions = ".".join(str(v) for v in self.env.version_info[:3])
+
+        content = Factory.create_pyproject_from_package(package=package)
+
+        for key in preserved:
+            content[key] = preserved[key]
+
+        self.system_pyproject.write_text(content.as_string(), encoding="utf-8")
+
+    def reset_poetry(self) -> None:
+        with directory(self.system_pyproject.parent):
+            self.generate_system_pyproject()
+            self._poetry = Factory().create_poetry(
+                self.system_pyproject.parent, io=self._io, disable_plugins=True
+            )
+
+    @property
+    def poetry(self) -> Poetry:
+        if self._poetry is None:
+            self.reset_poetry()
+        assert self._poetry
```

Comment:
Unnecessary, right?

Reply:
That was to satisfy mypy. I will replace that with `return cast(Poetry, self._poetry)` instead.